### PR TITLE
feat(plugin-docsearch): make docsearch async and only loads at click, close #1247

### LINKED ISSUE #1247: Make docsearch aync
### Clear and concise description of the problem

https://github.com/vuepress/vuepress-next/discussions/1242

### Suggested solution

Similar solution: https://github.com/vuejs/vitepress/blob/main/src/client/theme-default/components/VPNavBarSearch.vue

### Alternative

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Mister-Hope**

### CODE REVIEW

**Comment by meteorlxy** [comment]
It's better to split them into separated composables. Now there are too many lines in the component file.

**Comment by Mister-Hope** [comment]
Working on it now

**Comment by Mister-Hope** [comment]
Done

File: ecosystem/plugin-docsearch/src/client/components/Docsearch.ts [hunk 1]
```diff
@@ -1,14 +1,10 @@
...
 import type { DocsearchOptions } from '../../shared/index.js'
-import { useDocsearchShim } from '../composables/index.js'
+import { useDocsearch } from '../composables/index.js'
+import { options, searchButtonTemplate } from '../utils/index.js'
 
 declare const __DOCSEARCH_INJECT_STYLES__: boolean
-declare const __DOCSEARCH_OPTIONS__: DocsearchOptions
-const options = __DOCSEARCH_OPTIONS__
 
 if (__DOCSEARCH_INJECT_STYLES__) {
```
Comment (meteorlxy) [inline ecosystem/plugin-docsearch/src/client/components/Docsearch.ts:9]: duplicated?

Comment (Mister-Hope) [reply ecosystem/plugin-docsearch/src/client/components/Docsearch.ts:9]: It's a Boolean value, to ensure the code can be treeshaken, I would prefer keeping it.

`if (false)` can be removed all times, but not

```ts
const a = 1;

if (a) {
 // xxx
}
```

File: ecosystem/plugin-docsearch/src/client/composables/useDocSearch.ts [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+  const routeLocale = useRouteLocale()
+  const lang = usePageLang()
+  const docsearchShim = useDocsearchShim()
+
+  // to avoid loading the docsearch js upfront (which is more than 1/3 of the
+  // payload), we delay initializing it until the user has actually clicked or
+  // hit the hotkey to invoke it.
+  const loading = ref(false)
+  const loaded = ref(false)
+  const metaKey = ref(`'Meta'`)
```
Comment (meteorlxy) [inline ecosystem/plugin-docsearch/src/client/composables/useDocSearch.ts:23]: Seems that it's useless?

**Review by meteorlxy — APPROVED** [review]

**Pull request merged**